### PR TITLE
[assimp] update to 5.4.3 and fix build error

### LINKED ISSUE #40891: [assimp] update to 5.4.3
### Library name

assimp

### New version number

5.4.3

### Other information that may be useful (release notes, etc...)

Assimp 5.4.3 Bugfix Release

### PR DESCRIPTION
**Pull request by JonLiu1993**
Fixes #40891

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version.~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [ ] ~~Any patches that are no longer applied are deleted from the port's directory.~~
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

Tested usage successfully by assimp:x64-windows:
````
assimp provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(assimp CONFIG REQUIRED)
  target_link_libraries(main PRIVATE assimp::assimp)

assimp provides pkg-config modules:

  # Import various well-known 3D model formats in an uniform manner.
  assimp
````

### CODE REVIEW

File: ports/assimp/fix-incompatible.patch [hunk 1]
```diff
@@ -0,0 +1,26 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 80353ed..222a95f 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -316,7 +316,7 @@ ELSEIF(MSVC)
+     message("-- MSVC PDB generation disabled. Release binary will not be debuggable.")
+   endif()
+   # Source code is encoded in UTF-8
+-  ADD_COMPILE_OPTIONS(/source-charset:utf-8)
++  #ADD_COMPILE_OPTIONS(/source-charset:utf-8)
```
Comment (JonLiu1993) [inline ports/assimp/fix-incompatible.patch:10]: Fix error:
````
cl : Command line error D8016 : '/utf-8' and '/source-charset:utf-8' command-line options are incompatible
````

File: ports/assimp/fix-incompatible.patch [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+diff --git a/code/CMakeLists.txt b/code/CMakeLists.txt
+index a6db1cb..be3b731 100644
+--- a/code/CMakeLists.txt
++++ b/code/CMakeLists.txt
+@@ -1236,7 +1236,7 @@ ENDIF()
+ # stb
+ IF(1)
+   #hunter_add_package(stb)
+-  find_package(stb REQUIRED)
++  find_package(Stb REQUIRED)
```
Comment (JonLiu1993) [inline ports/assimp/fix-incompatible.patch:36]: Fix error:
````
CMake Error at /mnt/vcpkg-ci/installed/x64-linux/share/stb/vcpkg-cmake-wrapper.cmake:4 (_find_package):
  By not providing "Findstb.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "stb", but
  CMake did not find one.

  Could not find a package configuration file provided by "stb" with any of
  the following names:

    stbConfig.cmake
    stb-config.cmake
````

**Review by jimwang118 — DISMISSED** [review]

File: ports/assimp/build_fixes.patch [hunk 3]
```diff
@@ -76,10 +72,10 @@ index 6551dcb..457fcd7 100644
  @PACKAGE_INIT@
  
 +include(CMakeFindDependencyMacro)
-+
++find_dependency(kubazip CONFIG)
```
Comment (Osyotr) [inline ports/assimp/build_fixes.patch:75]: Why is this outside of `!shared` now?

File: ports/assimp/build_fixes.patch [hunk 4]
```diff
@@ -1,21 +1,17 @@
...
 +  #SET(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /D_DEBUG /Zi /Od")
-+  #SET(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE}")
-+  #SET(CMAKE_SHARED_LINKER_FLAGS_RELEASE "${CMAKE_SHARED_LINKER_FLAGS_RELEASE} /DEBUG:FULL /PDBALTPATH:%_PDB% /OPT:REF /OPT:ICF")
- ELSEIF (CMAKE_CXX_COMPILER_ID MATCHES "Clang" )
-   IF(NOT ASSIMP_HUNTER_ENABLED)
-     SET(CMAKE_POSITION_INDEPENDENT_CODE ON)
-@@ -421,7 +421,7 @@ ENDIF()
+   # Allow user to disable PDBs
+   if(ASSIMP_INSTALL_PDB)
+     SET(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /Zi")
```
Comment (Osyotr) [inline ports/assimp/build_fixes.patch:13]: You should probably set `-DASSIMP_INSTALL_PDB=OFF` and rely on `vcpkg_copy_pdbs`.

File: ports/assimp/build_fixes.patch [hunk 5]
```diff
@@ -76,10 +72,10 @@ index 6551dcb..457fcd7 100644
  @PACKAGE_INIT@
  
 +include(CMakeFindDependencyMacro)
-+
++find_dependency(kubazip CONFIG)
 +find_dependency(pugixml CONFIG)
```
Comment (Osyotr) [inline ports/assimp/build_fixes.patch:76]: I'm not sure if this still needs to be outside of `!shared`. The only place it's used in is `XmlParser.h` which isn't used by other public headers.

File: ports/assimp/build_fixes.patch [hunk 6]
```diff
@@ -399,7 +382,7 @@ index f4c2936..979e688 100644
 -  if(TARGET zip::zip)
 -    target_link_libraries(assimp PUBLIC zip::zip)
 +  if(1)
-+    target_link_libraries(assimp PRIVATE kubazip::kubazip)
++    target_link_libraries(assimp PUBLIC kubazip::kubazip)
```
Comment (Osyotr) [inline ports/assimp/build_fixes.patch:385]: ditto why is this `PUBLIC` now?

**Review by Osyotr — COMMENTED** [review]
Note:
This version adds vendored `tinyusdz` dependency (https://github.com/assimp/assimp/commit/0cb169368956fba4508fbd3af361cf37c8060e8e)

File: ports/assimp/build_fixes.patch (hunk 3 above)
Comment (JonLiu1993) [reply ports/assimp/build_fixes.patch:75]: When I tested usaege, I get this error, so I put it outside.
````
1> [CMake] CMake Error at F:/test/vcpkg/installed/x64-windows/share/assimp/assimpTargets.cmake:60 (set_target_properties):
1> [CMake]   The link interface of target "assimp::assimp" contains:
1> [CMake] 
1> [CMake]     kubazip::kubazip
````

File: ports/assimp/build_fixes.patch (hunk 4 above)
Comment (JonLiu1993) [reply ports/assimp/build_fixes.patch:13]: The portfile has been set to `-DASSIMP_INSTALL_PDB=OFF` and rely on `vcpkg_copy_pdbs`

File: ports/assimp/build_fixes.patch (hunk 3 above)
Comment (Osyotr) [reply ports/assimp/build_fixes.patch:75]: Well this is the reason: https://github.com/microsoft/vcpkg/pull/40894/files/c24c53b6d6e1da65e0a6e2198c0aa6fb850fdead#r1758951803

File: ports/assimp/build_fixes.patch (hunk 6 above)
Comment (JonLiu1993) [reply ports/assimp/build_fixes.patch:385]: This is a mistake of mine, I have changed it to PRIVATE.

File: ports/assimp/build_fixes.patch (hunk 5 above)
Comment (JonLiu1993) [reply ports/assimp/build_fixes.patch]: This fix uage error:
````
CMakeProject1\LINK : fatal error LNK1104: cannot open file 'pugixml.lib'
````

File: ports/assimp/build_fixes.patch (hunk 3 above)
Comment (JonLiu1993) [reply ports/assimp/build_fixes.patch]: I restored the code here.

File: ports/assimp/build_fixes.patch (hunk 5 above)
Comment (Osyotr) [reply ports/assimp/build_fixes.patch]: The reason is the same: the library is linked with PUBLIC keyword but IMHO it shouldn't.
https://github.com/microsoft/vcpkg/blob/86fba13068f1e8111e7d885de2aef11378435139/ports/assimp/build_fixes.patch#L388-L396

File: ports/assimp/fix-incompatible.patch (hunk 2 above)
Comment (Osyotr) [reply ports/assimp/fix-incompatible.patch:37]: This change was in a previous patch.

File: ports/assimp/fix-incompatible.patch [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+@@ -315,8 +315,6 @@ ELSEIF(MSVC)
+   elseif((GENERATOR_IS_MULTI_CONFIG) OR (CMAKE_BUILD_TYPE MATCHES Release))
+     message("-- MSVC PDB generation disabled. Release binary will not be debuggable.")
+   endif()
+-  # Source code is encoded in UTF-8
+-  ADD_COMPILE_OPTIONS(/source-charset:utf-8)
+ ELSEIF (CMAKE_CXX_COMPILER_ID MATCHES "Clang" )
+   IF(NOT ASSIMP_HUNTER_ENABLED)
+     SET(CMAKE_POSITION_INDEPENDENT_CODE ON)
+diff --git a/code/AssetLib/3MF/D3MFExporter.cpp b/code/AssetLib/3MF/D3MFExporter.cpp
```
[the remark this answers is not in the document]
Comment (JonLiu1993) [reply ports/assimp/fix-incompatible.patch]: Done

File: ports/assimp/build_fixes.patch (hunk 5 above)
Comment (JonLiu1993) [reply ports/assimp/build_fixes.patch]: Done

File: ports/assimp/fix-incompatible.patch (hunk 2 above)
Comment (JonLiu1993) [reply ports/assimp/fix-incompatible.patch]: Done

**Review by Osyotr — APPROVED** [review]

**Pull request merged**